### PR TITLE
feat(expr): add json path functions

### LINKED ISSUE #11372: Support JSON Path functions for JSONB
### Is your feature request related to a problem? Please describe.

TLDR; I want a dynamic processing for json columns.

I want to use RW to process events from many different clients, and to standardize the events. 
For example I have a "profile" json and I want to standardize out of it an "age" field.
For client A, he will have "age" field directly inside the json, for client B he will have "birthday" and I'll have to do "now - profile->birthday"

### Describe the solution you'd like

I imagine using [JMESPath](https://docs.rs/jmespath/latest/jmespath/) , and then having a mapping table which I can join against and then do (with some new '-@>' jmespath operator):

```
select profile-@>age_path as age
from events
join mappings on event.client_id = mappings.client_id
```

### Describe alternatives you've considered

Do the mapping somewhere else before reaching RW

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

resolve #11372 

This PR implements all jsonpath operators and functions:

```sql
jsonb @? jsonpath → boolean
jsonb @@ jsonpath → boolean
jsonb_path_exists ( target jsonb, path jsonpath [, vars jsonb [, silent boolean ]] ) → boolean
jsonb_path_match ( target jsonb, path jsonpath [, vars jsonb [, silent boolean ]] ) → boolean
jsonb_path_query ( target jsonb, path jsonpath [, vars jsonb [, silent boolean ]] ) → setof jsonb
jsonb_path_query_array ( target jsonb, path jsonpath [, vars jsonb [, silent boolean ]] ) → jsonb
jsonb_path_query_first ( target jsonb, path jsonpath [, vars jsonb [, silent boolean ]] ) → jsonb
```

This PR doesn't introduce the `jsonpath` data type. However, this does not affect the normal use of the above functions. All `jsonpath` arguments in the signature are actually varchar. If we have `jsonpath` type, there will be only one parsing during `'literal' :: jsonpath`. Currently, parsing happens every time the `eval` is called (not every time evaluating a row). I think it's not a big problem.

The jsonpath parsing and query logic is separated into an independent crate: [sql-json-path](https://github.com/risingwavelabs/sql-json-path). Please also review it if you are interested. (After investigation, I found no existing crate that can meet our requirements. Most of them follow the [IETF standard](https://datatracker.ietf.org/wg/jsonpath/about/) and only support serde-json.)

Almost all jsonpath syntax is supported, except `.datetime()` and some escaped string literal. I use Postgres regression test `jsonb_jsonpath.sql` to ensure correctness. Please check this file to see what is (not) supported yet. Detailed explanation: https://github.com/risingwavelabs/risingwave/pull/13568#issuecomment-1823966459

This PR is based on #13545, which is needed to run regression test. After it merged, I will rebase this PR onto main.

References:
- [Postgres Documentation for JSON Path](https://www.postgresql.org/docs/16/functions-json.html#FUNCTIONS-SQLJSON-PATH)
- [SQL/JSON Path standard manual](http://www.sai.msu.su/~megera/postgres/files/c067367_ISO_IEC_TR_19075-6_2017.pdf)
- [Gentle Guide to JSONPATH in PostgreSQL](https://github.com/obartunov/sqljsondoc/blob/master/jsonpath.md)
- [Databend's Implementation](https://github.com/datafuselabs/jsonb/tree/main/src/jsonpath)

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

Support jsonpath operators and functions:

```sql
jsonb @? jsonpath → boolean
jsonb @@ jsonpath → boolean
jsonb_path_exists ( target jsonb, path jsonpath [, vars jsonb [, silent boolean ]] ) → boolean
jsonb_path_match ( target jsonb, path jsonpath [, vars jsonb [, silent boolean ]] ) → boolean
jsonb_path_query ( target jsonb, path jsonpath [, vars jsonb [, silent boolean ]] ) → setof jsonb
jsonb_path_query_array ( target jsonb, path jsonpath [, vars jsonb [, silent boolean ]] ) → jsonb
jsonb_path_query_first ( target jsonb, path jsonpath [, vars jsonb [, silent boolean ]] ) → jsonb
```

See [postgres docs](https://www.postgresql.org/docs/current/functions-json.html) and search "jsonpath" for more details. Note that we don't support `jsonpath` data type for now. All `jsonpath` arguments in the signature are actually varchar. 

For the syntax of `jsonpath`, see [9.16.2. The SQL/JSON Path Language](https://www.postgresql.org/docs/current/functions-json.html#FUNCTIONS-SQLJSON-PATH). Almost all features in the docs have been supported, except that:

1. `.datetime()` and `. datetime(template)` method is not supported.
2. Some implementations for the `silent` argument may be incorrect.
3. Some implementations for regexp matching may be incorrect.

### CODE REVIEW

File: src/expr/impl/src/scalar/jsonb_path.rs [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+// limitations under the License.
+
+use jsonbb::ValueRef;
+use risingwave_common::types::{JsonbRef, JsonbVal};
+use risingwave_expr::{function, ExprError, Result};
+use sql_json_path::{EvalError, JsonPath, ParseError};
+
+#[function(
+    "jsonb_path_exists(jsonb, varchar) -> boolean",
+    prebuild = "JsonPath::new($1).map_err(parse_error)?"
```
Comment (BugenZhao) [inline src/expr/impl/src/scalar/jsonb_path.rs:22]: Great!

**Comment by wangrunji0408** [comment]
I have fixed regression tests as much as possible.
The remaining failing cases are:

```diff
 # wrong silent handling
 select jsonb_path_query('[1,"2",3]', '+$', '{}', true);
  jsonb_path_query 
 ------------------
- 1
-(1 row)
+(0 rows)

 # wrong silent handling
 select jsonb '["1",2,0,3]' @? '-$[*]';
  ?column? 
 ----------
- t
+ 
 (1 row)

 # double precision problem
 select jsonb_path_query('{"a": 2.5}', '-($.a * $.a).floor() % 4.3');
- jsonb_path_query 
-------------------
- -1.7
+  jsonb_path_query   
+---------------------
+ -1.7000000000000002
 (1 row)

 # I think both results are valid according to the specification. It depends on implementation.
 select jsonb_path_query('[[null, 1, "abc", "abcabc"]]', 'lax $ ? (@[*] starts with "abc")');
-      jsonb_path_query      
-----------------------------
- [null, 1, "abc", "abcabc"]
-(1 row)
+ jsonb_path_query 
+------------------
+(0 rows)

 # wrong regexp matching
 select jsonb_path_query('[null, 1, "a\b", "a\\b", "^a\\b$"]', 'lax $[*] ? (@ like_regex "a\\b" flag "")');
  jsonb_path_query 
 ------------------
+ "^a\\b$"
+ "a\\b"
  "a\b"
-(1 row)
+(3 rows)

 # wrong regexp matching
 select jsonb_path_query('[null, 1, "a\b", "a\\b", "^a\\b$"]', 'lax $[*] ? (@ like_regex "^a\\b$" flag "")');
  jsonb_path_query 
 ------------------
- "a\b"
-(1 row)
+(0 rows)

 # wrong silent handling
 SELECT jsonb_path_query_first('[{"a": 1}, {"a": 2}, {}]', 'strict $[*].a', '{}', true);
  jsonb_path_query_first 
 ------------------------
- 1
+ 
 (1 row)
```

File: src/expr/impl/Cargo.toml [hunk 2]
```diff
@@ -42,9 +42,13 @@ serde = { version = "1", features = ["derive"] }
 serde_json = "1"
 sha1 = "0.10"
 sha2 = "0.10"
+sql-json-path = { git = "https://github.com/risingwavelabs/sql-json-path.git", rev = "95a06aa", features = [
+    "jsonbb",
+] }
 thiserror = "1"
 tokio = { version = "0.2", package = "madsim-tokio", features = ["time"] }
 tracing = "0.1"
+# sql-json-path = { path = "../../../../sql-json-path", features = ["jsonbb"] }
```
Comment (TennyZhuang) [inline src/expr/impl/Cargo.toml:51]: Remove the comment?

File: src/expr/impl/src/scalar/jsonb_path.rs [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+)]
+fn jsonb_path_query2<'a>(
+    target: JsonbRef<'a>,
+    path: &JsonPath,
+) -> Result<impl Iterator<Item = JsonbVal> + 'a> {
+    let matched = path
+        .query::<ValueRef<'_>>(target.into())
+        .map_err(eval_error)?;
+    Ok(matched.into_iter().map(|json| json.into_owned().into()))
```
Comment (BugenZhao) [inline src/expr/impl/src/scalar/jsonb_path.rs:150]: Is it possible to return an iterator for `query`?

**Review by BugenZhao — COMMENTED** [review]
Cool!

**Review by TennyZhuang — APPROVED** [review]
The PR and test suite LGTM. Although not all corner cases were covered, I think it’s still good enough to merge.

File: src/expr/impl/src/scalar/jsonb_path.rs (hunk 3 above)
Comment (wangrunji0408) [reply src/expr/impl/src/scalar/jsonb_path.rs:150]: It is hard in stable Rust as all query functions need to be written as generator.

Comment (BugenZhao) [reply src/expr/impl/src/scalar/jsonb_path.rs:150]: What about those 3rd party libraries available under `stable` 🥵? Like https://github.com/Wolvereness/remit/

Comment (wangrunji0408) [reply src/expr/impl/src/scalar/jsonb_path.rs:150]: Interesting! Will have a try later.

**Pull request merged**